### DISCUSSION TITLE
Slow nonlinear convergence limit parameter

### DISCUSSION THREAD

**Discussion by am-tc01** [idaholab/moose#24457] [Q&A General]
Hi,
Is there any such parameter that defines the "slowness" of convergence of nonlinear solver i.e. if nonlinear residual is not decreasing at a considerable rate then just abort and cut the time-step in half, instead of waiting until number of iteration reaches `nl_max_its`. 
I googled a bit and found an [old discussion,](https://groups.google.com/g/moose-users/c/mn6nHwFMgPk/m/7Dfqsyz3dqsJ) which is exactly what I am looking for. I wonder if it was ever implemented?

Regards.

**Comment by GiudGiud** [comment]
Hello

you can do this with parsed expressions of postprocessors that contain the non linear residual. 

Then this expression of the ratio or a more complex expression will need to be fed into a Terminator (https://mooseframework.inl.gov/source/userobjects/Terminator.html) with a soft termination. This will just fail the time step without stopping the simulation

Guillaume

**Reply by am-tc01** [reply]
Hi @GiudGiud,

Yes that could do the job but not quite exactly. I tried using Terminator with `execute_on = LINEAR`, it does show the INFO message that `terminator is causing a time step cutback by marking the current step as failed` when the criteria is met but it still keeps on running that time step until the max. no of iterations are reached. What I want is to stop that time step as soon as the criteria is met and move on to the next step with half the time step.

**Reply by GiudGiud** [reply]
could you please paste the Terminator here?
I think we want fail_mode = SOFT and error-level of INFO ?

**Reply by am-tc01** [reply]
Yes, that's exactly what I used:
```
[terminator]
  type = Terminator
  expression = 'nonlinear_residual < 1E-5'
  fail_mode = SOFT
  error_level = INFO
  execute_on = 'LINEAR'
[]
```

**Reply by GiudGiud** [reply]
ok it seems we want a harder fail_mode here, but not "HARD" because that's going to error out the simulation.

I ll see if I can drum that up

**Reply by am-tc01** [reply]
Yes precisely!
Many thanks!

**Reply by GiudGiud** [reply]
it's a little bit of a pandora's box to throw somewhat uncontrollably within a solve. 
I made this branch, please feel free to try it out but I m not sure we ll be able to merge it very soon
https://github.com/idaholab/moose/pull/24463

**Reply by am-tc01** [reply]
Super!
I agree, it could be a bit tricky to throw uncontrollably, have to make sure it's safe. But I'll give it a shot.
Many thanks!!

Regards.

**Reply by GiudGiud** [reply]
ok found another solution without throwing. The PR is updated. I think we should be able to add that to MOOSE

**Reply by am-tc01** [reply]
Great! this sounds better than throwing. Thanks again!!
